File: lib/data/models/user_model.dart
class UserModel {
  final String id;
  final String email;
  final String name;
  final String? profilePicture;
  final String authProvider;
  final SubscriptionModel subscription;
  final PreferencesModel? preferences;
  final StatsModel? stats;
  final DateTime? createdAt;

  UserModel({
    required this.id,
    required this.email,
    required this.name,
    this.profilePicture,
    required this.authProvider,
    required this.subscription,
    this.preferences,
    this.stats,
    this.createdAt,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['id'] ?? json['_id'],
      email: json['email'],
      name: json['name'],
      profilePicture: json['profilePicture'],
      authProvider: json['authProvider'] ?? 'email',
      subscription: SubscriptionModel.fromJson(json['subscription'] ?? {}),
      preferences: json['preferences'] != null
          ? PreferencesModel.fromJson(json['preferences'])
          : null,
      stats: json['stats'] != null ? StatsModel.fromJson(json['stats']) : null,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'email': email,
      'name': name,
      'profilePicture': profilePicture,
      'authProvider': authProvider,
      'subscription': subscription.toJson(),
      'preferences': preferences?.toJson(),
      'stats': stats?.toJson(),
      'createdAt': createdAt?.toIso8601String(),
    };
  }

  bool get isPremium =>
      subscription.plan == 'premium' && subscription.status == 'active';

  UserModel copyWith({
    String? id,
    String? email,
    String? name,
    String? profilePicture,
    String? authProvider,
    SubscriptionModel? subscription,
    PreferencesModel? preferences,
    StatsModel? stats,
    DateTime? createdAt,
  }) {
    return UserModel(
      id: id ?? this.id,
      email: email ?? this.email,
      name: name ?? this.name,
      profilePicture: profilePicture ?? this.profilePicture,
      authProvider: authProvider ?? this.authProvider,
      subscription: subscription ?? this.subscription,
      preferences: preferences ?? this.preferences,
      stats: stats ?? this.stats,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

class SubscriptionModel {
  final String plan;
  final String status;
  final DateTime? startDate;
  final DateTime? endDate;

  SubscriptionModel({
    required this.plan,
    required this.status,
    this.startDate,
    this.endDate,
  });

  factory SubscriptionModel.fromJson(Map<String, dynamic> json) {
    return SubscriptionModel(
      plan: json['plan'] ?? 'free',
      status: json['status'] ?? 'active',
      startDate:
          json['startDate'] != null ? DateTime.parse(json['startDate']) : null,
      endDate: json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'plan': plan,
      'status': status,
      'startDate': startDate?.toIso8601String(),
      'endDate': endDate?.toIso8601String(),
    };
  }
}

class PreferencesModel {
  final bool notificationsEnabled;
  final ReminderModel meditationReminder;
  final ReminderModel moodReminder;

  PreferencesModel({
    required this.notificationsEnabled,
    required this.meditationReminder,
    required this.moodReminder,
  });

  factory PreferencesModel.fromJson(Map<String, dynamic> json) {
    return PreferencesModel(
      notificationsEnabled: json['notificationsEnabled'] ?? true,
      meditationReminder:
          ReminderModel.fromJson(json['meditationReminder'] ?? {}),
      moodReminder: ReminderModel.fromJson(json['moodReminder'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'notificationsEnabled': notificationsEnabled,
      'meditationReminder': meditationReminder.toJson(),
      'moodReminder': moodReminder.toJson(),
    };
  }
}

class ReminderModel {
  final bool enabled;
  final String time;

  ReminderModel({
    required this.enabled,
    required this.time,
  });

  factory ReminderModel.fromJson(Map<String, dynamic> json) {
    return ReminderModel(
      enabled: json['enabled'] ?? false,
      time: json['time'] ?? '09:00',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'enabled': enabled,
      'time': time,
    };
  }
}

class StatsModel {
  final int totalMeditations;
  final int totalMeditationTime;
  final String totalMeditationTimeFormatted;
  final int consecutiveDays;
  final DateTime? lastMeditationDate;
  final int journalEntries;
  final int journalEntriesThisMonth;
  final dynamic journalEntriesRemaining; // Can be int or 'unlimited'

  StatsModel({
    required this.totalMeditations,
    required this.totalMeditationTime,
    required this.totalMeditationTimeFormatted,
    required this.consecutiveDays,
    this.lastMeditationDate,
    required this.journalEntries,
    required this.journalEntriesThisMonth,
    this.journalEntriesRemaining,
  });

  factory StatsModel.fromJson(Map<String, dynamic> json) {
    return StatsModel(
      totalMeditations: json['totalMeditations'] ?? 0,
      totalMeditationTime: json['totalMeditationTime'] ?? 0,
      totalMeditationTimeFormatted:
          json['totalMeditationTimeFormatted'] ?? '0m',
      consecutiveDays: json['consecutiveDays'] ?? 0,
      lastMeditationDate: json['lastMeditationDate'] != null
          ? DateTime.parse(json['lastMeditationDate'])
          : null,
      journalEntries: json['journalEntries'] ?? 0,
      journalEntriesThisMonth: json['journalEntriesThisMonth'] ?? 0,
      journalEntriesRemaining: json['journalEntriesRemaining'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'totalMeditations': totalMeditations,
      'totalMeditationTime': totalMeditationTime,
      'totalMeditationTimeFormatted': totalMeditationTimeFormatted,
      'consecutiveDays': consecutiveDays,
      'lastMeditationDate': lastMeditationDate?.toIso8601String(),
      'journalEntries': journalEntries,
      'journalEntriesThisMonth': journalEntriesThisMonth,
      'journalEntriesRemaining': journalEntriesRemaining,
    };
  }
}

File: backend/src/models/User.model.js
const mongoose = require('mongoose');
const bcrypt = require('bcrypt');
const config = require('../config');

const userSchema = new mongoose.Schema({
  email: {
    type: String,
    required: [true, 'El correo electrónico es requerido'],
    unique: true,
    lowercase: true,
    trim: true,
    match: [/^\S+@\S+\.\S+$/, 'Formato de correo electrónico inválido'],
  },
  password: {
    type: String,
    minlength: [8, 'La contraseña debe tener al menos 8 caracteres'],
    select: false, // No incluir password en queries por defecto
  },
  name: {
    type: String,
    required: [true, 'El nombre es requerido'],
    trim: true,
  },
  profilePicture: {
    type: String,
    default: null,
  },
  authProvider: {
    type: String,
    enum: ['email', 'google'],
    default: 'email',
  },
  googleId: {
    type: String,
    unique: true,
    sparse: true, // Permite valores null sin violar la restricción unique
  },
  subscription: {
    plan: {
      type: String,
      enum: ['free', 'premium'],
      default: 'free',
    },
    status: {
      type: String,
      enum: ['active', 'canceled', 'past_due', 'trialing'],
      default: 'active',
    },
    startDate: Date,
    endDate: Date,
    stripeCustomerId: String,
    stripeSubscriptionId: String,
  },
  preferences: {
    notificationsEnabled: {
      type: Boolean,
      default: true,
    },
    meditationReminder: {
      enabled: {
        type: Boolean,
        default: false,
      },
      time: {
        type: String,
        default: '09:00',
      },
    },
    moodReminder: {
      enabled: {
        type: Boolean,
        default: false,
      },
      time: {
        type: String,
        default: '20:00',
      },
    },
  },
  stats: {
    totalMeditations: {
      type: Number,
      default: 0,
    },
    totalMeditationTime: {
      type: Number,
      default: 0, // En segundos
    },
    consecutiveDays: {
      type: Number,
      default: 0,
    },
    lastMeditationDate: Date,
    journalEntries: {
      type: Number,
      default: 0,
    },
    journalEntriesThisMonth: {
      type: Number,
      default: 0,
    },
    lastJournalReset: Date,
  },
  resetPasswordToken: String,
  resetPasswordExpire: Date,
  fcmTokens: [String], // Firebase Cloud Messaging tokens para notificaciones push
}, {
  timestamps: true,
});

// Index para búsquedas eficientes
userSchema.index({ email: 1 });
userSchema.index({ googleId: 1 });

// Hash password before saving
userSchema.pre('save', async function (next) {
  // Solo hashear si el password fue modificado o es nuevo
  if (!this.isModified('password')) {
    return next();
  }

  // Si authProvider es google, no hashear password
  if (this.authProvider === 'google') {
    return next();
  }

  try {
    const salt = await bcrypt.genSalt(config.bcryptRounds);
    this.password = await bcrypt.hash(this.password, salt);
    next();
  } catch (error) {
    next(error);
  }
});

// Method to compare passwords
userSchema.methods.comparePassword = async function (candidatePassword) {
  try {
    return await bcrypt.compare(candidatePassword, this.password);
  } catch (error) {
    throw new Error('Error al comparar contraseñas');
  }
};

// Method to check if user is premium
userSchema.methods.isPremium = function () {
  return this.subscription.plan === 'premium' &&
         this.subscription.status === 'active';
};

// Method to check if user can create journal entry
userSchema.methods.canCreateJournalEntry = function () {
  if (this.isPremium()) {
    return true;
  }

  // Free users: 10 entries per month
  return this.stats.journalEntriesThisMonth < 10;
};

// Method to reset monthly journal count
userSchema.methods.resetMonthlyJournalCount = function () {
  const now = new Date();
  const lastReset = this.stats.lastJournalReset;

  if (!lastReset ||
      lastReset.getMonth() !== now.getMonth() ||
      lastReset.getFullYear() !== now.getFullYear()) {
    this.stats.journalEntriesThisMonth = 0;
    this.stats.lastJournalReset = now;
  }
};

// Remove sensitive data from JSON output
userSchema.methods.toJSON = function () {
  const user = this.toObject();
  delete user.password;
  delete user.resetPasswordToken;
  delete user.resetPasswordExpire;
  delete user.__v;
  return user;
};

module.exports = mongoose.model('User', userSchema);
